packages/jupyter-ai/src/components/pending-messages.tsx

```diff
@@ -0,0 +1,141 @@
+import React, { useState, useEffect } from 'react';
+
+import { Box } from '@mui/material';
+import { AiService } from '../handler';
+import { ChatMessageHeader } from './chat-messages';
+
+const PENDING_MESSAGE_CLASS = 'jp-ai-pending-message';
+
+type PendingMessagesProps = {
+  messages: AiService.PendingMessage[];
+};
+
+type PendingMessageElementProps = {
+  text: string;
+  ellipsis: boolean;
+};
+
+type PendingMessageGroup = {
+  // Creating lastMessage as an AgentChatMessage
+  // as a hacky way to reuse ChatMessageHeader
+  // TODO: Refactor ChatMessageHeader to accept PendingMessage
+  lastMessage: AiService.AgentChatMessage;
+  messages: AiService.PendingMessage[];
+};
+
+function PendingMessageElement(props: PendingMessageElementProps): JSX.Element {
+  let text = props.text;
+  if (props.ellipsis) {
+    const [dots, setDots] = useState('');
+
+    useEffect(() => {
+      const interval = setInterval(() => {
+        setDots(dots => (dots.length < 3 ? dots + '.' : ''));
+      }, 500);
+
+      return () => clearInterval(interval);
+    }, []);
+    text = props.text + dots;
+  }
+  return (
+    <div>
+      {text.split('\n').map((line, index) => (
+        <p key={index}>{line}</p>
+      ))}
+    </div>
+  );
+}
+
+export function PendingMessages(props: PendingMessagesProps): JSX.Element {
+  if (props.messages.length === 0) {
+    return <></>;
+  }
+
+  const [timestamps, setTimestamps] = useState<Record<string, string>>({});
+  const personaGroups = groupMessages(props.messages);
+  /**
+   * Effect: update cached timestamp strings upon receiving a new message.
+   */
+  useEffect(() => {
+    const newTimestamps: Record<string, string> = { ...timestamps };
+    let timestampAdded = false;
+
+    for (const message of props.messages) {
+      if (!(message.id in newTimestamps)) {
+        // Use the browser's default locale
+        newTimestamps[message.id] = new Date(message.time * 1000) // Convert message time to milliseconds
+          .toLocaleTimeString([], {
+            hour: 'numeric', // Avoid leading zero for hours; we don't want "03:15 PM"
+            minute: '2-digit'
+          });
+
+        timestampAdded = true;
+      }
+    }
+    if (timestampAdded) {
+      setTimestamps(newTimestamps);
+    }
+  }, [personaGroups.map(group => group.lastMessage)]);
```

Comment:
1. Since we can assume there is only ever one persona, we can also assume there is only ever one timestamp that needs to be rendered. Given this, you can replace `timestamps: Record<string, string>` with a new state variable `timestamp: string`. 
2. This effect hook should depend on `props.messages` in the new implementation, as `personaGroups` can be dropped.